syntax = "proto3";

option java_multiple_files = true;
option java_package = "protobuf_experiments";
option java_outer_classname = "PBExperiments";
option objc_class_prefix = "PBE";

package protobuf_experiments;

message MessageHeader {
    uint32 version = 1;
    MessageType message_type = 2;
    string vin = 3;
}

enum MessageType {
    UNKNOWN = 0;
    AVAILABLE_PROJECTS_REQUEST = 1;
    AVAILABLE_PROJECTS_RESPONSE = 2;
    PROJECT_REQUEST = 3;
    PROJECT_RESPONSE = 4;
}

message Request {
    MessageHeader header = 1;
    oneof payload {
        AvailableProjectsRequest available_projects = 2;
        ProjectRequest project = 3;
    }
}

message AvailableProjectsRequest {
}

message ProjectRequest {
    string name = 1;
}

message Response {
    MessageHeader header = 1;
    oneof payload {
        AvailableProjectsResponse available_projects = 2;
        ProjectResponse project = 3;
    }
}

message AvailableProject {
    string name = 1;
    uint32 crc = 2;
}

message AvailableProjectsResponse {
    repeated AvailableProject available_projects = 1;
}

message ProjectResponse {
    bytes data = 1;
}
